### ISSUE TITLE
how to add middleware for api.ai?

### ISSUE THREAD

**Issue by vesper8** [mpociot/botman#102]
Hi there,

In your docs you show how to add a middleware for Wit.ai

```
$botman->middleware(Wit::create('MY-WIT-ACCESS-TOKEN'));
$botman->hears('emotion', function($bot) {
    $extras = $bot->getMessage()->getExtras();
    // Access extra information
    $entities = $extras['entities'];
});
```

And indeed you include the Wit middleware in your package.

But what about for Api.ai ? Are we supposed to create our own middleware?

Also it would be really helpful if you could add a fully working basic example that makes use of these middlewares

Btw am loving your package! Got set up with a facebook bot in 20 minutes flat!

**Comment by mpociot** [comment]
Hi!

Yes - right now you would need to write your own middleware to make use of api.ai.

I was talking to @christophrumpel about this and I know that he wanted to make a pull request with an api.ai middleware.

I know that the documentation still needs some improvements on all these little special areas like middlewares. I'm working on it :)

**Comment by christophrumpel** [comment]
I have already a working draft. I will make a PR today or tomorrow.

**Comment by vesper8** [comment]
@christophrumpel that's awesome!!

A working basic example would be super appreciated as well!

**Comment by christophrumpel** [comment]
No problem, I can share the working draft in about one hour.

**Comment by christophrumpel** [comment]
Faster than expected. Here you go:

Middleware: https://gist.github.com/christophrumpel/f7d4a18994c15ef7195219d9649c3c30

Use the middleware: (here I am replying with all the extras I have collected in the middleware which are the reply, the action and the intent (if given).

```php
        // create an instance
        $botman = app('botman');

        // give the bot something to listen for.
        $botman->middleware(ApiAi::create('your-api-ai-token'));
        $botman->hears('.*', function ($bot) {
            $extras = $bot->getMessage()->getExtras();
            $bot->reply(implode(' ,' , $extras));
        });

        // start listening
        $botman->listen();
```

**Comment by cerw** [comment]
@christophrumpel is this the offical way ".*" catching the whole input ? Or do we match the actual name of the  intents? ta

**Comment by christophrumpel** [comment]
Hey. My reply is quite old :-)

Here is the info to the new way: https://botman.io/2.0/nlp#api-ai

The best way is to only listen to API.ai actions like shown in the docs.

**Comment by nicola-rossato-opensymbol** [comment]
Meanwhile they have changed their company name :] so you need to change the middleware and the documentation consequentially :]

**Comment by christophrumpel** [comment]
Oh yeah I noticed... dont like the extra work for us but I am glad that the name is finally better :-) Guess we have to wait for another release because it is a breaking change, right @mpociot ?

**Comment by nicola-rossato-opensymbol** [comment]
@christophrumpel I'd like to ask you how to use the middleware for DialogFlow only when I need, for example I started a conversation, I return to the user a button with the action, when I process this action what I can do to trigger the middleware only for that message? I have built a class with the run method, and any action has it's method, some need to use the middleware and some don't. Thank you in advance for any suggestion.
